File: lernfeld_03_oop/cpp/loesungen/aufgabe_05.cpp
// Aufgabe 5: Objekt-Lebenszeiten und RAII — Musterlösung (C++)
//
// Im Gegensatz zu Python (Garbage Collection) ist die Zerstörung in C++
// deterministisch: Ein Objekt wird genau dann zerstört, wenn sein Scope
// ({ ... }) endet – in umgekehrter Reihenfolge der Erstellung. Das nennt
// man RAII (Resource Acquisition Is Initialization).
//
// Zusätzlich wird die Kopiersemantik sichtbar gemacht (Kopierkonstruktor).
// Wichtig: Wer eine der "Rule of Three" (Kopierkonstruktor, Kopierzuweisung,
// Destruktor) selbst schreibt, sollte alle drei im Blick haben.
//
// Kompilieren:  g++ -std=c++17 -Wall -Wextra aufgabe_05.cpp -o aufgabe_05
// Ausführen:    ./aufgabe_05

#include <iostream>  // std::cout, std::endl
#include <string>    // std::string
#include <vector>    // std::vector

class ProtokollObjekt {
private:
    std::string name;

public:
    // Konstruktor: wird bei der Erstellung aufgerufen
    ProtokollObjekt(const std::string& n) : name(n) {
        std::cout << name << " wird erstellt" << std::endl;
    }

    // Kopierkonstruktor – nur zur Veranschaulichung der Kopiersemantik.
    // (Für ein reines std::string-Member wäre die Compiler-Version perfekt.)
    ProtokollObjekt(const ProtokollObjekt& anderes) : name(anderes.name) {
        std::cout << name << " wird kopiert" << std::endl;
    }

    // Kopierzuweisung – gehört nach der "Rule of Three" dazu
    ProtokollObjekt& operator=(const ProtokollObjekt& anderes) {
        name = anderes.name;
        std::cout << name << " wird zugewiesen" << std::endl;
        return *this;
    }

    // Destruktor: wird beim Ende des Scopes aufgerufen
    ~ProtokollObjekt() {
        std::cout << name << " wird zerstoert" << std::endl;
    }

    // Vergleich der Namen (C++ kennt kein automatisches __eq__)
    bool gleicher_name(const ProtokollObjekt& anderes) const {
        return name == anderes.name;
    }
};

void funktion_mit_objekt() {
    ProtokollObjekt b("B");
    std::cout << "  (Funktion laeuft ...)" << std::endl;
    // b wird hier – am Funktionsende – zerstört
}

int main() {
    // Experiment 1: einzelnes Objekt auf dem Stack
    std::cout << "Experiment 1: einzelnes Objekt" << std::endl;
    {
        ProtokollObjekt a("A");
    }  // Scope zu Ende -> a wird zerstört

    // Experiment 2: Objekt in einer Funktion
    std::cout << "\nExperiment 2: Objekt in Funktion" << std::endl;
    funktion_mit_objekt();

    // Experiment 3: Objekte in einem std::vector
    std::cout << "\nExperiment 3: Objekte in Vektor" << std::endl;
    std::vector<ProtokollObjekt> liste;
    liste.reserve(2);  // Platz für beide vorab – ohne reserve würde der
                       // Vektor beim Wachsen die Elemente umkopieren
                       // (Reallokation, im Log als "wird kopiert" sichtbar!)
    liste.emplace_back("C");  // direkt im Vektor konstruiert (keine Kopie)
    liste.emplace_back("D");
    std::cout << "Liste wird geleert" << std::endl;
    liste.clear();  // garantiert: zuerst D, dann C (umgekehrte Reihenfolge)

    // Experiment 4: Vergleich der Namen
    std::cout << "\nExperiment 4: Vergleich" << std::endl;
    {
        ProtokollObjekt e("E");
        ProtokollObjekt f("E");  // gleicher Name wie e
        std::cout << "e.gleicher_name(f) ist "
                  << (e.gleicher_name(f) ? "true" : "false") << std::endl;
    }  // f, dann e werden zerstört

    // Experiment 5: Kopiersemantik sichtbar machen
    std::cout << "\nExperiment 5: Kopiersemantik" << std::endl;
    {
        ProtokollObjekt original("G");
        ProtokollObjekt kopie = original;  // Kopierkonstruktor!
    }  // kopie, dann original werden zerstört

    return 0;
}

File: lernfeld_03_oop/python/loesungen/aufgabe_05.py
"""Aufgabe 5: Objekt-Lebenszeiten und Dunder-Methoden — Musterlösung (Python).

Experiment-Programm: `__init__` und `__del__` machen den Lebenszyklus
sichtbar (CPython: Referenzzählung – ein Objekt wird zerstört, sobald die
letzte Referenz verschwindet). `__str__`, `__repr__` und `__eq__` steuern,
wie Objekte ausgegeben und verglichen werden.
"""


class ProtokollObjekt:
    """Ein Objekt, das jede Erstellung und Zerstörung meldet."""

    def __init__(self, name: str) -> None:
        self.name = name
        print(f"{name} wird erstellt")

    def __del__(self) -> None:
        print(f"{self.name} wird zerstört")

    def __str__(self) -> str:
        # Für Menschen lesbar – print(obj) und str(obj) nutzen das
        return self.name

    def __repr__(self) -> str:
        # Eindeutig und möglichst rekonstruierbar – repr(obj) nutzt das
        return f"ProtokollObjekt('{self.name}')"

    def __eq__(self, anderes) -> bool:
        # Zwei Objekte sind gleich, wenn ihre Namen gleich sind.
        # (Ohne __eq__ würde Python nur die Identität vergleichen.)
        if not isinstance(anderes, ProtokollObjekt):
            return NotImplemented
        return self.name == anderes.name


def funktion_mit_objekt() -> None:
    """Erzeugt ein Objekt – am Funktionsende verschwindet die Referenz."""
    b = ProtokollObjekt("B")
    print("  (Funktion läuft ...)")
    # b wird hier (am Funktionsende) zerstört


def main() -> None:
    # Experiment 1: einzelnes Objekt mit del löschen
    print("Experiment 1: einzelnes Objekt")
    a = ProtokollObjekt("A")
    del a  # letzte Referenz weg -> sofort zerstört

    # Experiment 2: Objekt in einer Funktion
    print("\nExperiment 2: Objekt in Funktion")
    funktion_mit_objekt()

    # Experiment 3: Objekte in einer Liste – die Liste leeren
    print("\nExperiment 3: Objekte in Liste")
    liste = [ProtokollObjekt("C"), ProtokollObjekt("D")]
    print("Liste wird geleert")
    liste.clear()  # C und D werden zerstört (CPython: zuletzt zuerst)

    # Experiment 4: Vergleich mit ==
    # Hinweis: Beide Objekte heißen "E" – so liefert __eq__ (Namensvergleich)
    # True. (In der Aufgabenstellung steht "F wird erstellt" – dort hieße das
    # zweite Objekt ebenfalls "E", damit der Vergleich True ergibt.)
    print("\nExperiment 4: Vergleich")
    e = ProtokollObjekt("E")
    f = ProtokollObjekt("E")  # gleicher Name wie e
    print(f"e == f ist {e == f}")  # True dank __eq__
    del f
    del e

    # Experiment 5: __str__ und __repr__ im Vergleich
    print("\nExperiment 5: __str__ und __repr__")
    g = ProtokollObjekt("G")
    print(f"print(g): {g}")  # nutzt __str__
    print(f"repr(g): {repr(g)}")  # nutzt __repr__
    del g


if __name__ == "__main__":
    main()
